pkg/events/base.go

```diff
@@ -0,0 +1,52 @@
+package events
+
+import (
+	"time"
+
+	"github.com/filecoin-project/lassie/pkg/types"
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multicodec"
+)
+
+type retrievalEvent struct {
+	eventTime   time.Time
+	retrievalId types.RetrievalID
+	payloadCid  cid.Cid
+}
+
+func (r retrievalEvent) Time() time.Time                { return r.eventTime }
+func (r retrievalEvent) RetrievalId() types.RetrievalID { return r.retrievalId }
+func (r retrievalEvent) PayloadCid() cid.Cid            { return r.payloadCid }
+
+type spRetrievalEvent struct {
+	retrievalEvent
+	storageProviderId peer.ID
+}
+
+func (e spRetrievalEvent) StorageProviderId() peer.ID { return e.storageProviderId }
```

Comment:
as per my comment in types.go, we could take this opportunity to switch to "ProviderID" here, shortening the name and ditching the "SP" acronym entirely through events, like `EventWithProviderID` for the interface below.

not a big deal, not a trivial change so I don't want to hold this PR up